pulsar-client-reactive-api/src/main/java/org/apache/pulsar/reactive/client/api/CorrelatedMessageSendingException.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pulsar.reactive.client.api;
+
+import org.reactivestreams.Publisher;
+
+/**
+ * A wrapper exception used by {@link ReactiveMessageSender#sendMany(Publisher)} and
+ * {@link ReactiveMessageSender#sendManyCorrelated(Publisher)}. The
+ * {@link #getCorrelationKey(Class)} method will return the correlation information to
+ * find out which message failed to be sent.
+ */
+public class CorrelatedMessageSendingException extends RuntimeException {
+
+	private final Object correlationKey;
+
+	public CorrelatedMessageSendingException(Throwable cause, Object correlationKey) {
+		super(cause);
+		this.correlationKey = correlationKey;
+	}
+
+	@Override
+	public Throwable fillInStackTrace() {
+		return this;
+	}
+
+	public <T> T getCorrelationKey(Class<T> correlationKeyType) {
```

Comment:
Is there a big value to have to provide the class here instead of returning Object and letting the user do the cast on the client side (which might even not be needed depending on the use case)?

Reply:
returning `Object` is very uncommon on modern Java APIs.  like the way how Reactor solves the ugliness of needing to cast in code.  In Reactor, there's [Context](https://projectreactor.io/docs/core/release/api/reactor/util/context/Context.html) and [ContextView](https://projectreactor.io/docs/core/release/api/reactor/util/context/ContextView.html) . Please check ContextView for examples.

Reply:
In this case, the need to pass the `Class` might be extra overhead and the signature could be simply `<T> T getCorrelationKey();`.